Synchronize shutdown in stanza adapter

michalpristas opened the pull request:
**Description:** 
This PR takes emitter-converter-receiver pipeline and synchronize shutdown in a sense that lower level of the pipeline needs to be fully finished before next one is shut down.

**Link to tracking Issue:** #31074

**Testing:** UT
run 
```
 go test -run ^TestShutdownFlush$ . -count 10000 -failfast
```

**Documentation:** in code comments

michalpristas committed:
Synchronize shutdown in stanza adapter

djaglowski commented on pkg/stanza/adapter/converter.go:
```diff
@@ -59,6 +59,9 @@ type Converter struct {
 
 	stopOnce sync.Once
+
+	// stopChan is an internal communication channel signalling stop was called
```
It looks like this is only to signal the converter to stop, so we could rename it `converterChan` to be a bit more clear.

djaglowski commented on pkg/stanza/adapter/receiver.go:
```diff
@@ -80,56 +81,40 @@ func (r *receiver) Start(ctx context.Context, host component.Host) error {
 		}
 	}
+
+	r.set.Logger.Debug("Receive loop stopped")
```
nit: Might as well be more specific here
```suggestion
	r.set.Logger.Debug("Emitter loop stopped")
```

djaglowski commented on pkg/stanza/operator/helper/emitter.go:
```diff
@@ -145,9 +145,17 @@ func (e *LogEmitter) flusher(ctx context.Context) {
 			}
 		case <-ctx.Done():
+			// flush currently batched entries
+			for {
+				oldBatch := e.makeNewBatch()
+				if len(oldBatch) == 0 {
+					break
+				}
+				e.flush(context.Background(), oldBatch)
+			}
```
IMO this is a bit more clear
```suggestion
			for oldBatch := e.makeNewBatch(); len(oldBatch) > 0; oldBatch = e.makeNewBatch() {
				e.flush(context.Background(), oldBatch)
			}
```

michalpristas replied:
i like it

djaglowski reviewed (commented):
These appear to be nice improvements. Ranging over channels is a lot cleaner. Having a channel / wg for each function is much more precise. 

One thing I'm wondering about is the use of `context.Background()`. I wonder if this indicates there are still issues to iron out. What is your take on this @michalpristas?

Also would really appreciate a review by @swiatekm if possible.

michalpristas committed:
Update pkg/stanza/adapter/receiver.go

Co-authored-by: Daniel Jaglowski <jaglows3@gmail.com>

michalpristas committed:
Update pkg/stanza/operator/helper/emitter.go

Co-authored-by: Daniel Jaglowski <jaglows3@gmail.com>

michalpristas committed:
converterChan

michalpristas commented:
with context.Background used i'm fine with how it is signalling `flush` that we don't care about the cancellation now, passing a cancelled context to it would not be ideal and flush is still able to be unblocked on `ctx.Done` in case called from public API. i get it's not nicest thing though

djaglowski commented:
Unless anyone else wants to review this, I think we just need to resolve the CI failures

swiatekm committed:
Fix deadlocked test

After the shutdown changes in previous commits, a receiver can only shut
down after its emitter was shut down first.

swiatekm commented on pkg/stanza/adapter/integration_test.go:
```diff
@@ -151,7 +151,10 @@ func TestEmitterToConsumer(t *testing.T) {
 	require.NoError(t, err)
-	defer func() { require.NoError(t, logsReceiver.Shutdown(context.Background())) }()
+	defer func() {
+		require.NoError(t, logsReceiver.emitter.Stop())
```
Note: This is now necessary, because the receiver is not allowed to stop until the emitter does. In an actual pipeline, the emitter itself is owned by the pipeline, and the pipeline is responsible for closing it. Here, we need to do it manually.

djaglowski commented:
Thanks again @michalpristas and @swiatekm. It'll be great to have this fix in place.

I'm happy to review a followup PR if there's any cleanup in mind.

andrzej-stencel commented on pkg/stanza/operator/helper/emitter.go:
```diff
@@ -145,9 +145,13 @@ func (e *LogEmitter) flusher(ctx context.Context) {
 			}
 		case <-ctx.Done():
+			// flush currently batched entries
+			for oldBatch := e.makeNewBatch(); len(oldBatch) > 0; oldBatch = e.makeNewBatch() {
```
Any specific scenario that warrants using a `for` instead of doing it just once? I would assume that at the point when `LogEmitter.Stop` is called the rest of the Stanza pipeline is stopped, so no new entries should be landing in the batch.

Not that the `for` hurts, it will work "more than" fine. :slightly_smiling_face:  I just wonder if I'm missing something.

swiatekm replied:
I think it's not necessary in practice, because other components shouldn't be calling `Process` after `Stop` was already called. And this is how the pipeline acts. @djaglowski am I wrong here?
